tao_compiler/mlir/disc/transforms/shape_utils.cc

```diff
@@ -0,0 +1,1743 @@
+    int64_t max_rank = clamp.max().getType().cast<RankedTensorType>().getRank();
+    if (min_rank != 0) {
+      mapShapeEqual(clamp.operand(), clamp.min());
+    }
+    if (max_rank != 0) {
+      mapShapeEqual(clamp.operand(), clamp.max());
+    }
+    Value result = cast<lmhlo::LmhloOp>(op).getResultBuffer();
+    mapShapeEqual(clamp.operand(), result);
+  } else if (auto select = dyn_cast<lmhlo::SelectOp>(op)) {
+    int64_t pred_rank =
+        select.pred().getType().cast<RankedTensorType>().getRank();
+    int64_t true_rank =
+        select.on_true().getType().cast<RankedTensorType>().getRank();
+    int64_t false_rank =
+        select.on_false().getType().cast<RankedTensorType>().getRank();
+    Value result = cast<lmhlo::LmhloOp>(op).getResultBuffer();
+    if (pred_rank != 0) {
+      mapShapeEqual(select.pred(), result);
+    }
+    if (true_rank != 0) {
+      mapShapeEqual(select.on_true(), result);
+    }
+    if (false_rank != 0) {
+      mapShapeEqual(select.on_false(), result);
+    }
+  } else if (isa<lmhlo::DynamicReshapeOp, lmhlo::ReshapeOp>(op)) {
+    Value input = op->getOperand(0);
+    // The last operand is the output memref by design
+    int64_t num_operand = op->getNumOperands();
+    Value output = op->getOperand(num_operand - 1);
+    mapValueEquivalent(input, output, valueWithSameElements_);
+  }
+
+  return success();
+}
+
+LogicalResult ShapeAnalysis::applyDimEqualWithDimValue() {
+  std::unordered_map<DimValue, SmallVector<SymbolDim*, 4>, DimValueHash>
+      dimValue2DimSymbols;
+  for (auto symbol_value : dimSymbol2DimValue_) {
+    auto dimSymbol = symbol_value.first;
+    DimValue valueRoot = getRootDimValue(symbol_value.second);
+    auto& dims = dimValue2DimSymbols[valueRoot];
+    dims.push_back(dimSymbol);
+  }
+  for (auto valueDimsPair : dimValue2DimSymbols) {
+    auto dims = valueDimsPair.second;
+    auto dim = dims[0];
+    for (int64_t i = 1; i < dims.size(); i++) {
+      if (failed(mapDimEqual(dim, dims[i]))) {
+        return failure();
+      }
+    }
+  }
+
+  return success();
+}
+
+LogicalResult ShapeAnalysis::buildRegionMulDecompose(Region* region) {
+  if (region->getBlocks().size() != 1) {
+    return region->getParentOp()->emitError(
+        "only single block region is supported");
+  }
+  for (Block& block : *region) {
+    if (failed(buildBlockMulDecompose(&block))) return failure();
+  }
+  return success();
+}
+
+LogicalResult ShapeAnalysis::buildBlockMulDecompose(Block* block) {
+  // mapping each op inside the block
+  WalkResult result = block->walk([&](Operation* op) {
+    if (auto muli = dyn_cast_or_null<arith::MulIOp>(op)) {
+      DimValue lhs(muli.getOperand(0));
+      DimValue rhs(muli.getOperand(1));
+      DimValue result(muli.getResult());
+      auto& decompose = dimValueMulDecompose_[result];
+      decompose.emplace_back(std::vector<DimValue>({lhs, rhs}));
+    } else if (isa<lmhlo::DynamicReshapeOp, lmhlo::ReshapeOp>(op)) {
```

Comment:
what's the essential difference between this phase and ShapeAnalysis::buildBlockDimValueMap()?

Reply:
Discussed offline. `DimValue` is to build the relationship between tensor/buffer's dim to a value. While this phase is to build multiplication relationship between dims.